### PR TITLE
Major release: .NET 10 upgrade, event-driven architecture, plugin system, Docker CI/CD, and comprehensive hardening

### PR DESCRIPTION
**Pull request by StoneyEagle**
This release represents a major architectural overhaul of NoMercy MediaServer, encompassing a full .NET 10 upgrade, a new event-driven architecture, a plugin system, a redesigned tray application, a guided setup flow, 
resilient boot/degraded mode support, comprehensive security hardening, critical bug fixes, and a complete test suite. The codebase has been significantly refactored for maintainability and correctness.

---
1. New Features

Event-Driven Architecture

- Create NoMercy.Events project with core event interfaces (EVT-01)
- Implement InMemoryEventBus with thread-safe pub/sub (EVT-02)
- Define all 17 domain event classes across 7 domains (EVT-04)
- Add event publishing to media scan pipeline (EVT-05)
- Add event publishing to encoding pipeline (EVT-06)
- Add event publishing to playback services (EVT-07)
- Migrate SignalR broadcasting to event handlers (EVT-09)
- Add event logging/audit middleware via decorator pattern (EVT-10)
- Rewrite LibraryFileWatcher as event-driven architecture
- Replace direct side effects with event-driven architecture across controllers and jobs
- Implement event-driven architecture for library, drive, and user permission change handlers
- Add functionality to scan for new library items and process them asynchronously
- Invalidate libraries query when first item is added to a library

Plugin System

- Create NoMercy.Plugins.Abstractions with all plugin interfaces (PLG-01)
- Implement PluginManager with AssemblyLoadContext loading (PLG-02)
- Add plugin manifest parsing and lifecycle state machine (PLG-03)
- Implement plugin DI integration with IPluginServiceRegistrator (PLG-05)
- Create plugin configuration system with JSON persistence (PLG-06)
- Create plugin repository system for remote discovery (PLG-07)
- Add plugin management API endpoints with DI registration (PLG-08)
- Create plugin project template with dotnet new support (PLG-09)

Setup Flow

- Create SetupState state machine for server setup lifecycle (SETUP-01)
- Add SetupServer with minimal HTTP Kestrel for setup mode (SETUP-02)
- Add embedded setup web UI with OAuth + device grant flow (SETUP-03)
- Fix OAuth PKCE flow for setup web UI with server-side code challenge (SETUP-04)
- Enhance /sso-callback with OAuth error handling, token verification, and styled HTML (SETUP-05)
- Add SSE support to /setup/status endpoint for real-time progress updates (SETUP-06)
- Add 503 middleware for non-setup routes during setup mode (SETUP-07)
- Refactor Auth.cs — remove all Console.* calls for headless/service mode (SETUP-08)
- Add exponential backoff retry to Register.Init() and Certificate.RenewSslCertificate() (SETUP-09)
- Implement graceful HTTP -> HTTPS Kestrel restart after cert acquisition (SETUP-11)
- Replace Environment.Exit(1) in ApiInfo with network-first cache fallback (SETUP-12)
- Add end-to-end setup flow tests covering full install -> HTTPS lifecycle (SETUP-17)

Tray Application

- Create NoMercy.Tray Avalonia project (HEAD-08)
- Implement system tray icon with status indicator (HEAD-09)
- Implement log viewer window in tray application (HEAD-10)
- Implement server control UI in tray application (HEAD-11)
- Unify tray windows into single tabbed MainWindow
- Add server process launching from tray, fix log viewer and setup state
- Add real-time log streaming and startup transparency for Tray
- Add server name, port, and worker settings to launcher tray via IPC

Service Hosting & IPC

- Create management API controller with localhost-only access (HEAD-03)
- Implement named pipe IPC for local communication (HEAD-04)
- Add Windows Service hosting support (HEAD-05)
- Add Linux systemd and macOS launchd service support (HEAD-06)
- Manage App as child process of the Service
- Add auto-start on boot toggle with management API and tray UI

CLI

- Create NoMercy.Cli project with CLI commands (HEAD-13)
- Add start, resources, autostart, and update CLI commands
- Match CLI log output to server console format with SSE streaming

Resilient Boot / Degraded Mode

- Implement degraded mode startup so server boots without network (BOOT-01)
- Fix Cloudflare single point of failure with cache fallbacks (BOOT-02)
- Implement offline token validation with cached Keycloak public key (BOOT-03)
- Implement declarative startup dependency reordering (BOOT-04)
- Implement health check endpoints with DB and degraded mode status (BOOT-05)

Queue System Decoupling

- Create NoMercy.Queue.Core with interfaces and POCO models (QDC-01)
- Refactor JobQueue to accept IQueueContext via adapter pattern (QDC-05)
- Refactor JobDispatcher from static to instance class (QDC-06)
- Refactor QueueRunner from static to instance class (QDC-08)
- Create NoMercy.Queue.Sqlite standalone SQLite queue provider (QDC-10)
- Create NoMercy.Queue.MediaServer media-server queue infrastructure (QDC-13)
- Decouple queue system into standalone NoMercyQueue libraries
- Move RescanFilesJob to its own configurable "file" worker queue

Media Processing

- Add "Episode XX" filename parsing and TMDB episode group support for absolute numbering
- Implement directory matching logic to enhance folder path resolution across multiple managers
- Remove FFMpegCore dependency and use own ffprobe implementation
- Add global ffprobe concurrency throttle to prevent process explosion

Docker & Deployment

- Add Docker image builds (CPU, NVIDIA, Intel QSV, AMD) to CI/CD pipeline with ghcr.io
- Runtime-only Dockerfiles using pre-built self-contained binaries
- Docker Compose files updated to reference registry images

---
2. Performance Improvements

- Replace client-side search filtering with DB-side custom SQLite function (CRIT-02)
- Split heavy Include chains into focused queries (CRIT-03)
- Replace sync Shell.ExecStdOutSync with async parallel probing in HLS playlist generator (CRIT-12)
- Fix IQueryable returns — rename browsable methods, materialize terminal queries (HIGH-01)
- Add explicit [Index] annotations for missing FK indexes (HIGH-04)
- Add per-endpoint response caching with Cache-Control headers (HIGH-05)
- Throttle encoder progress updates to max 2/sec (HIGH-08)
- Add 500MB size limit and LRU pruning to dev cache (HIGH-11)
- Split GetContinueWatchingAsync into two-step query to avoid loading duplicate entity trees (MED-01)
- Replace .Count > 0 with .Any() for SQL EXISTS instead of COUNT (MED-02)
- Create ForUser() extension to eliminate repeated auth filtering (MED-03)
- Cache Localizer per language to avoid XML re-parsing on every request (MED-11)
- Use [GeneratedRegex] source generators to avoid runtime regex compilation in FfMpeg (MED-12)
- Parallelize independent startup tasks with Task.WhenAll (MED-16)
- Configure MemoryCache with SizeLimit and compaction (MED-20)
- Replace Include-based entity loading with Select projections and parallelize music start queries
- Reduce ffprobe concurrency and add exponential backoff for resource errors
- Build Windows/macOS/Linux installers from shared runtime instead of single-file exes

---
3. Security Hardening

- Gate EnableSensitiveDataLogging on Config.IsDev (HIGH-03)
- Fix middleware ordering — conditional DeveloperExceptionPage, reorder CORS/compression (HIGH-06)
- Gate SignalR EnableDetailedErrors on Config.IsDev (HIGH-07)
- Set finite Kestrel limits instead of unlimited (HIGH-14)
- Reduce SignalR MaximumReceiveMessageSize from 100MB to 2MB (MED-07)
- Remove hardcoded client secret, load config from env vars (MED-17)
- Make CORS dev origins conditional on Config.IsDev (MED-18)
- Remove client_secret from auth flows, add PKCE and behavioral tests
- Fix inverted expiration check in Auth.Init() (AUTH-BUG)

---
4. Bug Fixes

Critical

- Replace all new MediaContext() with DI-injected contexts (CRIT-01)
- Replace .Wait()/.Result deadlock patterns with async/sync alternatives (CRIT-04)
- Remove static MediaContext from ClaimsPrincipleExtensions, add Initialize method (CRIT-05)
- Replace lock(Context) with dedicated _writeLock object in JobQueue (CRIT-06)
- Implement IHttpClientFactory for all providers to prevent socket exhaustion (CRIT-07)
- Add exception handling and lifecycle tracking to QueueRunner worker threads (CRIT-08)
- Add missing return in ReserveJob recursive retry (CRIT-09)
- Add try-finally cleanup and ConcurrentDictionary to FFmpeg process management (CRIT-11)
- Remove duplicate DbContext registration that shadowed scoped with transient (CRIT-13)
- Reject non-IShouldQueue types in QueueWorker instead of silently hanging (CRIT-14)
- Replace .Result deadlock with proper await in TvdbBaseClient.GetToken (PROV-CRIT-03)
- Invert client-key condition in FanArtBaseClient so header is sent when key exists (PROV-CRIT-04)

High

- Remove | true debug bypass in ImageController that skipped all image processing (HIGH-15)
- Clean up partial output files on encoding failure (HIGH-09)
- Add exception logging and back-off to async void queue processor (HIGH-10)
- Add lock synchronization to QueueRunner worker counter (HIGH-16)
- Add ChangeTracker.Clear() after all SaveChanges in JobQueue (HIGH-17)
- Reduce retry attempts and add jitter to JobQueue retries (HIGH-18)
- Remove blocking .Result from ExternalIp property getter (HIGH-20)
- Remove all GC.Collect band-aids across 16 files (HIGH-20b)
- Remove duplicate cron job DI registration and add worker dedup guard (HIGH-13)
- Wrap Ffprobe.Kill() in try-catch for process termination race (HIGH-19)
- Handle port-in-use error gracefully instead of crashing (BOOT-06)

Provider / Model Fixes

- Remove dead contradictory while loop in AcoustIdBaseClient.Get (PROV-H06)
- Read response content once as byte[] in 3 image download clients (PROV-H10, H12, H16)
- Change MusixMatch AlbumName type from long to string? (PMOD-CRIT-01)
- Add missing setters to TVDB ForSeries and Tag properties (PMOD-CRIT-02)
- Change TMDB Video property type from string? to bool? (PMOD-CRIT-03)

Database Model Fixes

- Change Track.MetadataId type from int? to Ulid? (DBMOD-CRIT-01)
- Realign shifted JsonProperty names in Library.cs (DBMOD-CRIT-02)
- Add MaxLength(4096) to QueueJob.Payload and large-text fields (DBMOD-CRIT-03)
- Remove = new() initializers from nullable navigation properties in Cast.cs (DBMOD-CRIT-04)
- Change UserData.TvId JsonProperty from "episode_id" to "tv_id" (DBMOD-H01)
- Change duplicate JsonProperty("id") to "network_tv" on Network.NetworkTv (DBMOD-H02)

Disposal / Resource Management

- Add missing using/dispose to Image<Rgba32> in hot paths (DISP-01)
- Add missing using/dispose to HttpResponseMessage across 11 sites (DISP-02)
- Add missing using/dispose to TagLib.File / TagFile factory (DISP-03)
- Add missing using/dispose to cold path resources (DISP-04)

Miscellaneous

- Swap macOS cloudflared architecture download URLs (SYS-H14)
- Resolve DbContext concurrency in HomeService.GetHomeData()
- Replace fire-and-forget SSE writes with Channel-based consumer loop
- Actually renew SSL certificate when expiring within 30 days
- Parse show title from folder when filename starts with S##E## pattern
- Correct Special have_items count and update MCU timeline order
- Translate missing Dutch (nl) entries in I18N.xml
- Guard against ObjectDisposedException in ConnectivityManager.Dispose

---
5. Code Quality & Refactoring

Namespace / Folder Reorganization

- Rename services/ to Services/ for PascalCase namespace convention (REORG-01)
- Move MusicHub to Hubs/, music services to Services/Music/ (REORG-02)
- Move VideoHub to Hubs/, video services to Services/Video/ (REORG-03)
- Consolidate all API DTOs into NoMercy.Api/DTOs/ (REORG-04)
- Remove duplicate FolderDto and FolderLibraryDto classes (REORG-05)
- Organize 97 database models into domain subfolders (REORG-06)
- Remove empty NoMercy.EncoderV2 stub directories (REORG-07)
- Rename AppConfig/ to Configuration/ (REORG-09)
- Create centralized Extensions/ folders per project (REORG-10)
- Move Swagger config to Configuration/Swagger/ folder (REORG-11)

Architectural Refactoring

- Rename NoMercy.Server to NoMercy.Service and update CI/CD packaging
- Decompose monolithic Networking class into DI-friendly services with connectivity state machine
- Standardize repository method names and improve delete functionality across media controllers
- Rename jobs to {Subject}{Action}{Type} convention and split queues
- Reduce startup log verbosity and move server banner to end of startup
- Remove redundant logging in RemoveWatched method of VideoHub
- Remove dead System.Text.Json config, consolidate to Newtonsoft serializer (MED-08)
- Code quality cleanup — async naming, string allocation, catch blocks, image perf (LOW-01-10)
- Resolve all 77 build warnings to achieve zero-warning build
- Remove unused using statements across the codebase

---
6. Testing

Characterization Test Suite

- Add NoMercy.Tests.Api project with WebApplicationFactory and auth helpers (CHAR-01)
- Add NoMercy.Tests.Repositories project with in-memory SQLite and seed data (CHAR-02)
- Add snapshot tests for all /api/v1/ Media endpoints (CHAR-03)
- Add snapshot tests for all /api/v1/ Music endpoints (CHAR-04)
- Add snapshot tests for all /api/v1/ Dashboard endpoints (CHAR-05)
- Add query output tests for all repository methods via SQL interceptor (CHAR-06)
- Add SignalR hub connection tests for all 5 mapped hubs (CHAR-07)
- Add queue behavior tests for implicit retry, cross-queue isolation, and lifecycle (CHAR-08)
- Add encoder command-building tests capturing FFmpeg CLI args (CHAR-09)
- Add CI pipeline for characterization tests (CHAR-10)

Other Tests

- Add regression test for duplicate RequestLocalization removal (MED-19)
- Unify TMDB test collection to prevent API rate limit flakes
- Resolve all test failures across 9 root causes
- Add comprehensive queue testing and fix FindFailedJob int/long bug (QDC-17)

---
7. CI/CD & DevOps

- Add Docker image build/push to CI/CD pipeline with ghcr.io (CPU, NVIDIA, Intel QSV, AMD)
- Use pre-built self-contained binaries for Docker images (no .NET SDK in containers)
- Download only release artifacts in create-release workflow (skip Docker buildx cache)
- Add staged draft releases from non-dev branches
- Skip tests in staged release workflow (tests run on CI/CD pipeline to dev)
- Mark coverage job as non-blocking in test workflow
- Build Windows installer on native Windows runner instead of Wine
- Replace softprops/action-gh-release with gh CLI for reliability

---
8. Framework & Dependencies

- Upgrade entire solution from .NET 9.0 to .NET 10.0
- Create Directory.Packages.props with all 91 package versions (CPM-01)
- Remove Version attributes from all .csproj files and activate Central Package Management (CPM-02)
- Remove FFMpegCore dependency in favor of own ffprobe implementation

### CODE REVIEW

File: src/NoMercy.Api/Services/Music/MusicPlaylistManager.cs [hunk 1]
```diff
@@ -1,162 +1,162 @@
...
+        if (track is null)
+            throw new("Track not found");
+
+        // Load TrackUser data for favorite status
+        bool isFavorite = await _mediaContext.TrackUser
+            .AnyAsync(tu => tu.TrackId == trackId && tu.UserId == userId);
+        
+        if (isFavorite && !track.TrackUser.Any(tu => tu.UserId == userId))
+        {
+            track.TrackUser.Add(new() { TrackId = trackId, UserId = userId });
```
[the remark this answers is not in the document]
Comment (StoneyEagle) [reply src/NoMercy.Api/Services/Music/MusicPlaylistManager.cs:61]: @copilot how is it not obvious when i am adding a TrackUser?
